### PR TITLE
`IteratorEnvironment` fixes

### LINKED ISSUE #4810: RFileScanner's IteratorEnvironment throws UnsupportedOperationException for getConfig() and getServiceEnv()
**Describe the bug**
When attaching an iterator to an RFileScanner which uses either the iterator environments config or ServiceEnvironment, the iterators will throw UnsupportedOperationExceptions 

**Versions (OS, Maven, Java, and others, as appropriate):**
Accumulo 2.x

**To Reproduce**
Steps to reproduce the behavior (or a link to an example repository that reproduces the problem):
1. Create an RFileScanner
2. Attach an Iterator which in the init method calls env.getConfig() or env.getServiceEnv()

**Expected behavior**
The config for the RFileScanner should be built from any table properties passed into the builder's withTableProperties() method. These properties can also be returned for any env.getServiceEnv().getConfiguration() or env.getServiceEnv().getConfiguration(TableId) methods.

### PR DESCRIPTION
**Pull request by kevinrr888**
closes #4810
- Fixed some inconsistencies and bugs in (non-test) implementations of `IteratorEnvironment`
	- Fixed `RFileScanner`s unexpected `UnsupportedOperationException` when calling `env.getServiceEnv()` or `env.getConfig()` (both deprecated) or `env.getPluginEnv()`
	- Some impls returned a new `ServiceEnvironment` for each call to `getServiceEnv()`/`getPluginEnv()`. This did not align as well with the `IteratorEnvIT` which would assert that these two calls return the same object. Memoize a supplier to the `ServiceEnvironment` to avoid creating the object unless needed and to return the same object each time.
	- Seems like `isUserCompaction()` and `isFullMajorCompaction()` are supposed to be throwing an `IllegalStateException` if it's not a major compaction. Added missing javadoc to `isUserCompaction()` and changed impls that were not adhering to these docs.
- Added tests for these environments to `IteratorEnvIT`

Questions:
1) Need confirmation, but it was my interpretation that a TableId doesn't make sense in the context of `RFileScanner`. Is this correct? _**ANSWERED**_
2) I have `getConfiguration()` throw an error in `RFileScanner`s `IteratorEnvironment` since I'm not sure how we would return the system config in this context and I don't think it would even make sense here. Does this seem right? _**ANSWERED**_
3) Made the `IteratorEnvironment` for `RFileScanner` return null and not throw an error on `env.getTableId()` so something like: `env.getPluginEnv().getConfiguration(env.getTableId())` can be done. Thoughts on this? _**ANSWERED**_

### CODE REVIEW

File: core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java [hunk 1]
```diff
@@ -335,6 +352,52 @@ public boolean isSamplingEnabled() {
     public SamplerConfiguration getSamplerConfiguration() {
       return RFileScanner.this.getSamplerConfiguration();
     }
+
+    @Override
+    public TableId getTableId() {
+      return null;
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java:358]: Returning null could cause an NPE in calling code that could be much later than this call and if the NPE is in a compound statement it could be hard to work back to the cause.  Thinking it would be better to fail fast rather than return null.

```suggestion
      throw new UnsupportedOperationException("This scanner is unrelated to any table or accumulo instance, it is operating directly on files."):
```


Alternatively we could always return a dummy table id here.

```suggestion
     return TableId.of("rfileScannerFakeD");
```

If we did that then the methods that take a table id could possibly only accept this id.

File: core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java [hunk 2]
```diff
@@ -335,6 +352,52 @@ public boolean isSamplingEnabled() {
...
+        }
+
+        @Override
+        public <T> T instantiate(String className, Class<T> base) {
+          throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public String getTableName(TableId tableId) {
+          throw new UnsupportedOperationException();
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java:387]: All of these exceptions could include the same message.  Maybe make the message a private constant.

```suggestion
          throw new UnsupportedOperationException("This scanner is unrelated to any table or accumulo instance, it is operating directly on files. Therefore it can not support this operation.");
```

File: core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java [hunk 3]
```diff
@@ -335,6 +352,52 @@ public boolean isSamplingEnabled() {
...
+    private ServiceEnvironment createServiceEnv() {
+      return new ServiceEnvironment() {
+        @Override
+        public <T> T instantiate(TableId tableId, String className, Class<T> base) {
+          throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public <T> T instantiate(String className, Class<T> base) {
+          throw new UnsupportedOperationException();
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java:382]: Could implement this in a similar way to 

https://github.com/apache/accumulo/blob/fd111a22a6454837ad6076d5ba6fe54fc8934be5/core/src/main/java/org/apache/accumulo/core/clientImpl/ClientServiceEnvironmentImpl.java#L65-L69

File: core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java [hunk 4]
```diff
@@ -335,6 +352,52 @@ public boolean isSamplingEnabled() {
...
+        }
+
+        @Override
+        public String getTableName(TableId tableId) {
+          throw new UnsupportedOperationException();
+        }
+
+        @Override
+        public Configuration getConfiguration(TableId tableId) {
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java:391]: It would be nice to update the javadoc for the following methods.

https://github.com/apache/accumulo/blob/fd111a22a6454837ad6076d5ba6fe54fc8934be5/core/src/main/java/org/apache/accumulo/core/client/rfile/RFile.java#L178-L187

To state that the iterator env will return the properties passed into these methods.  Could also include the version number in the docs.  Like `Starting with 2.1.4 <link to iter env method> will return the properties passed in here`

**Review by keith-turner — COMMENTED** [review]
Still looking at this, posting comments I have so far.  Will circle back to this later.

File: core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java (hunk 1 above)
Comment (kevinrr888) [reply core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java]: Went with the latter, I assumed the way this would probably be used is `env.getPluginEnv().getConfiguration(env.getTableId())`. If we throw an error on `env.getTableId()` that might be confusing because not sure what user would pass to `getConfiguration(TableId)`... User could create their own dummy id but I feel it makes more sense for us to make it.

File: core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java (hunk 2 above)
Comment (kevinrr888) [reply core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java]: Done

File: core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java (hunk 3 above)
Comment (kevinrr888) [reply core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java]: Done

File: core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java (hunk 4 above)
Comment (kevinrr888) [reply core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java:399]: Added and fixed an issue I noticed with a couple RFile javadocs

File: core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java [hunk 5]
```diff
@@ -88,6 +95,7 @@ class RFileScanner extends ScannerOptions implements Scanner {
   private long readaheadThreshold = 3;
   private AccumuloConfiguration tableConf;
   private CryptoService cryptoService;
+  private final TableId dummyTableId;
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java:98]: This could be static

File: core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java [hunk 6]
```diff
@@ -335,6 +357,57 @@ public boolean isSamplingEnabled() {
...
+        @Override
+        public <T> T instantiate(String className, Class<T> base)
+            throws ReflectiveOperationException {
+          return this.getClass().getClassLoader().loadClass(className).asSubclass(base)
+              .getDeclaredConstructor().newInstance();
+        }
+
+        @Override
+        public String getTableName(TableId tableId) {
+          throw new UnsupportedOperationException(errorMsg);
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java:395]:
```suggestion
      if(!tableId.equals(dummyTableId)){
          throw new UnsupportedOperationException(errorMsg);
      } else {
          return "RFileScannerFakeTableName";
      }
```

File: test/src/main/java/org/apache/accumulo/test/IteratorEnvIT.java [hunk 7]
```diff
@@ -222,6 +243,49 @@ private void testScan(String tableName,
...
+      Class<? extends SortedKeyValueIterator<Key,Value>> iteratorClass) throws Exception {
+    writeData(tableName);
+    IteratorSetting is = new IteratorSetting(1, iteratorClass);
+    is.addOption("expected.table.id",
+        getServerContext().tableOperations().tableIdMap().get(tableName));
+
+    try (Scanner scanner = client.createScanner(tableName);
+        var clientIterScanner = new ClientSideIteratorScanner(scanner)) {
+      clientIterScanner.addScanIterator(is);
+      var unused = clientIterScanner.iterator();
```
Comment (keith-turner) [inline test/src/main/java/org/apache/accumulo/test/IteratorEnvIT.java:285]: This may be a preexisting problem with this test, not sure.  It seems like there is no positive confirmation that the init code was actually called.  If there was a bug and the init method on the iterator was never called then the checks that happen in init would never be done.  Once possible way to solve this is to filter data based on config passed to init.  Then if we read the data and see it was filtered we know the init method was called.  If this is a preexisting problem with the test we can open a follow on issue if that seems cleaner.

File: core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java (hunk 5 above)
Comment (kevinrr888) [reply core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java]: Done

File: core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java (hunk 6 above)
Comment (kevinrr888) [reply core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java]: Instead of throwing an UOE, I throw an IllegalArgumentException. Let me know if this approach is ok. Figured this might be a bit clearer from a user perspective

File: test/src/main/java/org/apache/accumulo/test/IteratorEnvIT.java (hunk 7 above)
Comment (kevinrr888) [reply test/src/main/java/org/apache/accumulo/test/IteratorEnvIT.java]: Yeah, good point. This seems to go in line with the current changes since I'm already changing the test quite a bit; seems fine to do in this PR. I refactored the test to ensure the expected checks (from `testEnv()`) are done for each test

File: core/src/main/java/org/apache/accumulo/core/iterators/IteratorEnvironment.java [hunk 8]
```diff
@@ -140,7 +140,8 @@ default SamplerConfiguration getSamplerConfiguration() {
   }
 
   /**
-   * True if compaction was user initiated.
+   * True if compaction was user initiated. Will throw IllegalStateException if
+   * {@link #getIteratorScope()} != {@link IteratorScope#majc}.
```
Comment (keith-turner) [inline core/src/main/java/org/apache/accumulo/core/iterators/IteratorEnvironment.java:144]:
```suggestion
   * @return True if compaction was user initiated, false otherwise.
   * @throws IllegalStateException if {@link #getIteratorScope()} != {@link IteratorScope#majc}
```

**Review by keith-turner — APPROVED** [review]

File: core/src/main/java/org/apache/accumulo/core/client/rfile/RFile.java [hunk 9]
```diff
@@ -172,17 +175,27 @@ public interface ScannerOptions {
      * {@link Property#TABLE_PREFIX} may be accepted and used. For example, cache and crypto
      * properties could be passed here.
      *
+     * <p>
+     * Starting with {@code 2.1.4}, {@link PluginEnvironment#getConfiguration(TableId)} (obtained by
+     * {@link IteratorEnvironment#getPluginEnv()}) will return the properties passed in here.
+     *
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/client/rfile/RFile.java:181]: I'm not sure these comments about "starting with 2.1.4" make sense to include here. It should have always worked this way. This is just fixing a bug. I don't think we need to document the implementation details of the corrected behavior for the bug fix in the javadoc.

File: core/src/main/java/org/apache/accumulo/core/client/ClientSideIteratorScanner.java [hunk 10]
```diff
@@ -134,15 +142,22 @@ public SamplerConfiguration getSamplerConfiguration() {
...
-      return new ClientServiceEnvironmentImpl(context.get());
+    @Deprecated(since = "2.0.0")
+    public AccumuloConfiguration getConfig() {
+      var ctx = context.get();
+      try {
+        return new ConfigurationCopy(
+            ctx.tableOperations().getConfiguration(ctx.getTableName(tableId.get())));
+      } catch (AccumuloException | TableNotFoundException e) {
+        throw new RuntimeException("Error getting table configuration", e);
+      }
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/client/ClientSideIteratorScanner.java:154]: What is the rationale for implementing `getConfig`, and specifically as a copy of the table configuration? The behavior of this method in the interface was never well-defined, and was using an internal API and should never be called by user code. I'm not sure why it needs an implementation here, rather than just throwing an exception, like the deprecated default method does in the interface.

If we really do need this, the default method in the interface should implement it by calling the methods on `getPluginEnv()` and wrapping it similarly to what you did here, instead of that default method throwing an exception.

File: core/src/main/java/org/apache/accumulo/core/client/ClientSideIteratorScanner.java [hunk 11]
```diff
@@ -134,15 +142,22 @@ public SamplerConfiguration getSamplerConfiguration() {
...
     }
 
+    @Deprecated(since = "2.1.0")
     @Override
-    public PluginEnvironment getPluginEnv() {
-      return new ClientServiceEnvironmentImpl(context.get());
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/client/ClientSideIteratorScanner.java:145]: This is the method that should be implemented. The `getServiceEnv()` method only existed for backwards compatibility. The code should not rely on the interface implementation of `getPluginEnv()` calling the `getServiceEnv()` here, because that behavior is removed when the deprecated code is dropped. Instead, it's best to keep both methods, as before your change, so that when the deprecated one is dropped in subsequent versions, the correct implementation is kept.

File: core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java [hunk 12]
```diff
@@ -66,18 +67,25 @@
...
+import com.google.common.base.Suppliers;
 
 class RFileScanner extends ScannerOptions implements Scanner {
 
   private static final byte[] EMPTY_BYTES = new byte[0];
   private static final Range EMPTY_RANGE = new Range();
+  private static final String errorMsg =
+      "This scanner is unrelated to any table or accumulo instance;"
+          + " it operates directly on files. Therefore, it can not support this operation.";
+  private static final TableId dummyTableId = TableId.of("RFileScannerFakeTableId");
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java:88]: Why wouldn't this be null?

File: core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java [hunk 13]
```diff
@@ -311,14 +319,27 @@ public void updateScanIteratorOption(String iteratorName, String key, String val
   }
 
   private class IterEnv implements IteratorEnvironment {
+    private final Supplier<ServiceEnvironment> serviceEnvironment;
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java:322]: What prevents us from just using the implementation from `super.getPluginEnv()` and `super.getServiceEnv()`?

File: core/src/main/java/org/apache/accumulo/core/iterators/IteratorEnvironment.java [hunk 14]
```diff
@@ -52,16 +52,16 @@ default AccumuloConfiguration getConfig() {
...
    */
   default IteratorScope getIteratorScope() {
     throw new UnsupportedOperationException();
   }
 
   /**
-   * Return true if the compaction is a full major compaction. Will throw IllegalStateException if
-   * {@link #getIteratorScope()} != {@link IteratorScope#majc}.
+   * @return true if the compaction is a full major compaction; false otherwise
+   * @throws IllegalStateException if {@link #getIteratorScope()} != {@link IteratorScope#majc}.
```
Comment (ctubbsii) [inline core/src/main/java/org/apache/accumulo/core/iterators/IteratorEnvironment.java:64]: I think I would have preferred fixing the documentation to drop the exception case, rather than forcing an exception to be thrown in all the scan cases. `false otherwise` should be sufficient.

File: core/src/main/java/org/apache/accumulo/core/client/rfile/RFile.java (hunk 9 above)
Comment (kevinrr888) [reply core/src/main/java/org/apache/accumulo/core/client/rfile/RFile.java:181]: Okay, I will remove this in my next commit. Should I just remove the mention of `2.1.4` or that new part entirely?

File: core/src/main/java/org/apache/accumulo/core/client/ClientSideIteratorScanner.java (hunk 10 above)
Comment (kevinrr888) [reply core/src/main/java/org/apache/accumulo/core/client/ClientSideIteratorScanner.java:154]: I implemented it here since I thought that because it was deprecated but not yet removed, there should still be functionality for the method. The other impls of `IteratorEnvironment` still have this implemented; assumed same should be the case here. Less importantly, this also made testing the different impls of `IteratorEnvironment` in `IteratorEnvIT` more straight forward as each would call `getConfig()`, `getServiceEnv()`, and `getPluginEnv()`. If this shouldn't be implemented here, I can remove it, or I could change the default impl to use `getPluginEnv()` like you said. Whichever you think would be better.

File: core/src/main/java/org/apache/accumulo/core/client/ClientSideIteratorScanner.java (hunk 11 above)
Comment (kevinrr888) [reply core/src/main/java/org/apache/accumulo/core/client/ClientSideIteratorScanner.java:145]: Thanks, this was something I was unsure about when implementing this. Some `IteratorEnvironment`s had both `getServiceEnv()` and `getPluginEnv()` (e.g., `OfflineIterator`) while some had just `getServiceEnv()` (e.g., `TabletIteratorEnvironment`), and I wasn't sure which to go with. I'll fix the impls to use both.

File: core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java (hunk 12 above)
Comment (kevinrr888) [reply core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java:88]: Discussed here: https://github.com/apache/accumulo/pull/4816#discussion_r1725159871. Either null or a dummy would work though.

File: core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java (hunk 13 above)
Comment (kevinrr888) [reply core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java:322]: `IteratorEnvironment` doesn't have any implementation for `getServiceEnv()` or `getPluginEnv()`. Or are you suggesting changing `IteratorEnvironment`?

File: core/src/main/java/org/apache/accumulo/core/iterators/IteratorEnvironment.java (hunk 14 above)
Comment (kevinrr888) [reply core/src/main/java/org/apache/accumulo/core/iterators/IteratorEnvironment.java:64]: I hadn't looked into why an exception was thrown, I had just kept the functionality the same and consistent across impls. I assumed there was probably a reason for throwing the exception and didn't look any further. 

Are you suggesting to drop the documentation about the exception case and revert the changes I made to adhere to the exception case, or are you suggesting changing the functionality to no longer throw the exception, or something else?

**Comment by ctubbsii** [comment]
After a conversation with @kevinrr888, I added commit 5924b02c065297886ede479fbcf9750e699e4362 to incorporate some of the changes we discussed, that he intends to build on and polish up. The main points included were:

* Push a reasonable `getConfig()` implementation up into `IteratorEnvironment`'s default method (depends on the ability to have `getTableId()` return a null value)
* Include concrete implementations of `getPluginEnv()` instead of just `getServiceEnv()` to make it easier to just drop the deprecated one when merging to newer branches
* Consider using a Proxy instance of TableId instead of a real instance, so it isn't possible to be confused with any similar user-constructed TableId (may not work)
* Update javadocs in RFile to make it clearer how properties set on the scanner builder (such as iterators) will be available to client-side iterators from the plugin environment

The main thing left unresolved after our conversation was how to fix the mismatch between the implementations and the javadoc for all the methods that inconsistently return null or throw IllegalStateException or UnsupportedOperationException, when calling a method that doesn't apply to a given situation (like asking for a table ID or table name when operating directly on RFiles or asking if it's a user-initiated compaction when it's a scan). Nearly every caller seems to have a unique set a requirements/implementation, and it's a bit hard to home in on the best implementation, especially for the default methods, and make sure the javadoc matches the behavior.

**Comment by kevinrr888** [comment]
### Comment about `IteratorEnvironment.getConfig()`:

It might be strange to have a default impl for this method but not others... If we do want a default impl for this that isnt just throwing a UOE, we should probably change the others too... But that should be done in a follow on if that is desired. Keeping the default as throwing a UOE keeps it consistent with the other default impls/more inline with the original approach (throw UOE by default, concrete impls should have an actual impl). The only `IteratorEnvironment` which would be using the default `getConfig()` would be `ClientSideIteratorScanner`. But we can just add the impl there (done in latest commit). It is also hard to know what would be a good default impl since its not exactly clear how `getConfig()` was originally working.

### Comment about `RFileScanner.IterEnv.getTableId()`

Proxy can only be used to create proxies for interfaces, not concrete classes, so can't be used for `TableId`. Opting to just return null with added javadoc.

Background: we want to support `getConfiguration(TableId)` since this is the method defined as returning the props set on the table and `getConfiguration()` should not be returning table props, so we cant substitute that method. So, something needs to be passed, but it doesn't matter what since it wont be used. The way `getConfiguration(TableId)` will be called is `env.getPluginEnv().getConfiguration(env.getTableId())` (from javadocs for `IteratorEnvironment.getPluginEnv()`) So, we should have `env.getTableId()` returning something. 

No solutions seem great, but `Proxy` is a no-go, a real dummy id may be confusing (e.g., a user may think that it is a real table id), null may lead to NPE. I figure null makes the most sense here. A NPE would only occur if the user is trying to do something with the table id which they shouldnt be doing anyways. It is also now noted in the javadocs that null will be returned so if a NPE does occur, it should be easy to see why.

### Comment about UOE/ISE

Did not change anything further than I had already changed in my earlier commits relating to when/how UOE or ISE are thrown, how its documented etc. Some future considerations would be
- `IteratorEnvironment.getAuthorizations()` currently documented as throwing `UnsupportedOperationException` if scope != scan (should probably be ISE if anything).
- `IteratorEnvironment.isFullMajorCompaction()` and `isUserCompaction()` should maybe throw ISE in the default if the scope is majc
	- Maybe these shouldn't be throwing exceptions at all and only returning false... This would probably be the cleanest if possible. Would have to look more into why these are throwing exceptions in the first place.

### Comment about `RFileScanner`s `createServiceEnv()`

Christopher suggested having this be an inner class instead of anon class. Leaving this comment as a reminder. Will do once the other more important changes are looked at; want to avoid confusing diff between commits for now.

**Comment by kevinrr888** [comment]
@ctubbsii - Could you give this another look over when you get the chance? Would be good to get this merged in.

File: core/src/main/java/org/apache/accumulo/core/iterators/IteratorEnvironment.java (hunk 14 above)
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/iterators/IteratorEnvironment.java:64]: I was suggesting just having the boolean methods return true or false... rather than throw an exception at all in some subset of "false" cases.

File: core/src/main/java/org/apache/accumulo/core/client/rfile/RFile.java (hunk 9 above)
Comment (ctubbsii) [reply core/src/main/java/org/apache/accumulo/core/client/rfile/RFile.java]: I think it could be dropped entirely.

**Comment by ctubbsii** [comment]
> @ctubbsii - Could you give this another look over when you get the chance? Would be good to get this merged in.

It's been long enough that I'm probably going to have to look at this from scratch... I don't remember all the nuances that led to my comments on my earlier reviews. I will add this to my list of items to set aside and review ASAP.

**Comment by dlmarion** [comment]
I'm coming late to this conversation, and I have not read all of the prior comments. I don't understand the design of the IteratorEnvironment interface, such that all of the methods have default implementations that throw an UnsupportedOperationException. If this has been addressed elsewhere, please point me to it and I'm happy to read through it.

This type of interface design pattern enables implementations to not have to handle all of the methods, which might make it easier on the developer. But it makes it easier to run into a runtime exception instead of a compilation error. I think I'd rather see the fix for this remove the default exception throwing method implementations on the interface. I'm happy to start a different PR for this if that's out of scope for this.

**Comment by kevinrr888** [comment]
@dlmarion - This was preexisting functionality added in https://github.com/apache/accumulo/pull/955/, and may be explained there. I think I went through this PR a while ago, but don't remember anything about it. Skimming through some of the old comments in this PR, it looks like I address some of the difficulties with this current impl, as you mentioned.

> Comment about IteratorEnvironment.getConfig():
TLDR: opted for not having a default impl for IteratorEnvironment.getConfig() (keep default as throwing UOE)
It might be strange to have a default impl for this method but not others... If we do want a default impl for this that isnt just throwing a UOE, we should probably change the others too... But that should be done in a follow on if that is desired. Keeping the default as throwing a UOE keeps it consistent with the other default impls/more inline with the original approach (throw UOE by default, concrete impls should have an actual impl). The only IteratorEnvironment which would be using the default getConfig() would be ClientSideIteratorScanner. But we can just add the impl there (done in latest commit). It is also hard to know what would be a good default impl since its not exactly clear how getConfig() was originally working.

So seems like changing IteratorEnvironment completely in this PR might be a scope creep (at least that's I thought 7 months ago)

**Comment by kevinrr888** [comment]
@dlmarion - While I don't remember much about this, your suggestion sounds like a good change to me, as the current design doesn't seem great to me either. 

If we want an UOE for some of the ops, we can define that in the concrete impls. This forces the developer to consider all of the methods individually and prevents the dev from forgetting to add certain functionality.

**Comment by dlmarion** [comment]
I have created #5490 as an alternate solution for #4810.

File: core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java [hunk 15]
```diff
@@ -335,6 +355,70 @@ public boolean isSamplingEnabled() {
...
+        }
+
+        @Override
+        public String getTableName(TableId tableId) {
+          throw new UnsupportedOperationException(errorMsg);
+        }
+
+        @Override
+        public Configuration getConfiguration(TableId tableId) {
+          Preconditions.checkArgument(tableId == getTableId(),
```
Comment (dlmarion) [inline core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java:410]: `getTableId` above only returns `null`.

Comment (kevinrr888) [reply core/src/main/java/org/apache/accumulo/core/client/rfile/RFileScanner.java:410]: Yes, the intention was that the only valid use of this method is:
`env.getPluginEnv().getConfiguration(env.getTableId())`
(as described in javadocs for `getPluginEnv()`)
`env.getPluginEnv().getConfiguration(null)`
would also work, and this still makes sense in the context of `RFileScanner`

Just wanted to avoid accepting something like
`env.getPluginEnv().getConfiguration(someRandomTableId)`

**Comment by kevinrr888** [comment]
https://github.com/apache/accumulo/pull/5490 closed https://github.com/apache/accumulo/issues/4810
Closing this PR; will open a new PR with changes we still want from this PR
